File: src/Crawler/Analysis/Result/SecurityResult.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Analysis\Result;

class SecurityResult
{

    /**
     * @var SecurityCheckedHeader[]
     */
    public array $checkedHeaders = [];

    public function getCheckedHeader(string $header): SecurityCheckedHeader
    {
        if (!isset($this->checkedHeaders[$header])) {
            $this->checkedHeaders[$header] = new SecurityCheckedHeader($header);
        }
        return $this->checkedHeaders[$header];
    }

    public function getHighestSeverity(): int
    {
        $highestSeverity = SecurityCheckedHeader::OK;
        foreach ($this->checkedHeaders as $item) {
            if ($item->highestSeverity > $highestSeverity) {
                $highestSeverity = $item->highestSeverity;
            }
        }
        return $highestSeverity;
    }

}

File: src/analysis/result/security_result.rs
// SiteOne Crawler - SecurityResult
// (c) Jan Reges <jan.reges@siteone.cz>

use indexmap::IndexMap;

use super::security_checked_header::{SEVERITY_OK, SecurityCheckedHeader};

#[derive(Debug, Clone, Default)]
pub struct SecurityResult {
    pub checked_headers: IndexMap<String, SecurityCheckedHeader>,
}

impl SecurityResult {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn get_checked_header(&mut self, header: &str) -> &mut SecurityCheckedHeader {
        self.checked_headers
            .entry(header.to_string())
            .or_insert_with(|| SecurityCheckedHeader::new(header.to_string()))
    }

    pub fn get_highest_severity(&self) -> i32 {
        let mut highest = SEVERITY_OK;
        for item in self.checked_headers.values() {
            if let Some(sev) = item.highest_severity
                && sev > highest
            {
                highest = sev;
            }
        }
        highest
    }
}
